spark3/src/main/java/org/apache/iceberg/actions/Spark3SnapshotAction.java

```diff
@@ -0,0 +1,148 @@
+import org.apache.iceberg.Table;
+import org.apache.iceberg.TableProperties;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.spark.SparkTableUtil;
+import org.apache.iceberg.spark.source.SparkTable;
+import org.apache.spark.sql.SparkSession;
+import org.apache.spark.sql.catalyst.TableIdentifier;
+import org.apache.spark.sql.catalyst.analysis.NoSuchNamespaceException;
+import org.apache.spark.sql.catalyst.analysis.TableAlreadyExistsException;
+import org.apache.spark.sql.connector.catalog.CatalogPlugin;
+import org.apache.spark.sql.connector.catalog.Identifier;
+import org.apache.spark.sql.connector.catalog.StagedTable;
+import org.apache.spark.sql.connector.catalog.TableCatalog;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import scala.collection.JavaConverters;
+
+/**
+ * Creates a new Iceberg table based on a source Spark table. The new Iceberg table will
+ * have a different data and metadata directory allowing it to exist independently of the
+ * source table.
+ */
+class Spark3SnapshotAction extends Spark3CreateAction implements SnapshotAction {
+  private static final Logger LOG = LoggerFactory.getLogger(Spark3SnapshotAction.class);
+
+  private String destTableLocation = null;
+
+  Spark3SnapshotAction(SparkSession spark, CatalogPlugin sourceCatalog,
+                       Identifier sourceTableName, CatalogPlugin destCatalog,
+                       Identifier destTableName) {
+    super(spark, sourceCatalog, sourceTableName, destCatalog, destTableName);
+  }
+
+  @Override
+  public Long execute() {
+    StagedTable stagedTable;
+    Table icebergTable;
+
+    try {
+      stagedTable = destCatalog().stageCreate(destTableName(), v1SourceTable().schema(),
+          sourcePartitionSpec(), buildPropertyMap());
+      icebergTable = ((SparkTable) stagedTable).table();
+
+      if (!icebergTable.properties().containsKey(TableProperties.DEFAULT_NAME_MAPPING)) {
+        assignDefaultTableNameMapping(icebergTable);
+      }
+    } catch (TableAlreadyExistsException taeException) {
+      throw new IllegalArgumentException("Cannot create snapshot because a table already exists with that name",
+          taeException);
+    } catch (NoSuchNamespaceException nsnException) {
+      throw new IllegalArgumentException("Cannot create snapshot because the namespace given does not exist",
+          nsnException);
+    }
+
+    boolean threw = true;
+    try {
+      String stagingLocation = icebergTable.location() + "/" + ICEBERG_METADATA_FOLDER;
+      LOG.info("Beginning snapshot of {} to {} using metadata location {}", sourceTableName(), destTableName(),
+          stagingLocation);
+
+      TableIdentifier v1TableIdentifier = v1SourceTable().identifier();
+      SparkTableUtil.importSparkTable(spark(), v1TableIdentifier, icebergTable, stagingLocation);
+      stagedTable.commitStagedChanges();
+      threw = false;
+    } finally {
+      if (threw) {
```

Comment:
Why cannot we just do the rollback in a catch clause? This seems a bit too complicated to me.

Reply:
This was recommend by @rdblue to insure the original exceptions are propagated as is while still doing cleanup